src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -310,9 +407,33 @@ private static bool RequiresRefOrOut(BindValueKind kind)
 
 #nullable enable
 
+        private static AccessorKind GetIndexerAccessorKind(BoundIndexerAccess indexerAccess, BindValueKind valueKind)
+        {
+            var coreValueKind = valueKind & ValueKindSignificantBitsMask;
+            var returnsByRef = indexerAccess.Indexer.RefKind != RefKind.None;
+            AccessorKind kind = (returnsByRef, coreValueKind) switch
```

Comment:
The `returnsByRef` has to figure into the logic. Consider this example: 

```csharp
S local;
local[0]++;

struct S {
    public ref int this[int index] => throw null!;
}
```

This will come across with `coreValueKind` of `BindValueKind.CompoundAssignment`. The only way to know if it uses both accessors or just `get` is to look at the `returnsByRef`. Can simplify this logic a bit though to only consider it in that case.

Reply:
> The `returnsByRef` has to figure into the logic.

To clarify, I am not suggesting to remove the check for `returnsByRef`. I am suggesting to perform it outside of the `switch`, i.e. before it. Something like:
```
            If (returnsByRef)
            {
                return AccessorKind.Get;
            }

            AccessorKind kind = coreValueKind switch
            {
                BindValueKind.CompoundAssignment => AccessorKind.Both,
                BindValueKind.Assignable => AccessorKind.Set,
                BindValueKind.RValue => AccessorKind.Get,

                // These combinations of not returns by ref but having a ref accessor kind 
                // occur in error scenarios.
                BindValueKind.RefOrOut => AccessorKind.Get,
                BindValueKind.RefAssignable => AccessorKind.Get,
                BindValueKind.ReadonlyRef => AccessorKind.Get,
                _ => AccessorKind.Unknown,
            };

            Debug.Assert(kind != AccessorKind.Unknown);
            return kind;
```